docker-qgis/qfc_worker/process_projectfile.py

```diff
@@ -99,47 +125,66 @@ def on_project_read(doc):
-
-        for i in range(nodes.size()):
-            node = nodes.item(i)
-            element = node.toElement()
-            if (
-                element.hasAttribute("name")
-                and element.attribute("name") == "theMapCanvas"
-            ):
-                map_settings.readXml(node)
-
-        map_settings.setRotation(0)
-        map_settings.setTransformContext(project.transformContext())
-        map_settings.setPathResolver(project.pathResolver())
-        map_settings.setOutputSize(QSize(100, 100))
-        map_settings.setLayers(reversed(list(layer_tree.customLayerOrder())))
-        # print(f'output size: {map_settings.outputSize().width()} {map_settings.outputSize().height()}')
-        # print(f'layers: {[layer.name() for layer in map_settings.layers()]}')
-
-    project.readProject.connect(on_project_read)
-    project.read(project.fileName())
+
+    def on_project_read_wrapper(
+        tmp_project: QgsProject,
+        tmp_layer_tree: QgsLayerTree,
+    ) -> Callable[[QDomDocument], None]:
+        def on_project_read(doc: QDomDocument) -> None:
+            r, _success = tmp_project.readNumEntry("Gui", "/CanvasColorRedPart", 255)
+            g, _success = tmp_project.readNumEntry("Gui", "/CanvasColorGreenPart", 255)
+            b, _success = tmp_project.readNumEntry("Gui", "/CanvasColorBluePart", 255)
+            map_settings.setBackgroundColor(QColor(r, g, b))
+
+            nodes = doc.elementsByTagName("mapcanvas")
+
+            for i in range(nodes.size()):
+                node = nodes.item(i)
+                element = node.toElement()
+                if (
+                    element.hasAttribute("name")
+                    and element.attribute("name") == "theMapCanvas"
+                ):
+                    map_settings.readXml(node)
+
+            map_settings.setRotation(0)
+            map_settings.setTransformContext(tmp_project.transformContext())
+            map_settings.setPathResolver(tmp_project.pathResolver())
+            map_settings.setOutputSize(QSize(100, 100))
+            map_settings.setLayers(reversed(list(tmp_layer_tree.customLayerOrder())))
+
+        return on_project_read
+
+    # NOTE use a temporary project to generate the layer rendering with improved speed
+    tmp_project = QgsProject()
+    tmp_layer_tree = tmp_project.layerTreeRoot()
+    tmp_project_read_flags = (
+        # TODO we use `QgsProject` read flags, as the ones in `Qgis.ProjectReadFlags` do not work in QGIS 3.34.2
+        QgsProject.ReadFlags()
+        | QgsProject.ForceReadOnlyLayers
+        | QgsProject.FlagDontLoadLayouts
+        | QgsProject.FlagDontLoad3DViews
+        | QgsProject.DontLoadProjectStyles
+    )
+    tmp_project.readProject.connect(
+        on_project_read_wrapper(tmp_project, tmp_layer_tree)
+    )
+    tmp_project.read(
+        project_filename,
+        tmp_project_read_flags,
+    )
 
     renderer = QgsMapRendererParallelJob(map_settings)
```

Comment:
Instead of the QEventLoop stuff, I'd just use renderer.waitForFinished()

Reply:
We did it in another way, see the last commit.